storage/cache/subtree_cache.go

```diff
@@ -211,28 +211,20 @@ func (s *SubtreeCache) preload(ids []storage.NodeID, getSubtrees GetSubtreesFunc
 		close(workTokens)
 	}()
 
-	// Note: It's only below this point where the subtrees map needs the lock.
-	s.mutex.Lock()
-	defer s.mutex.Unlock()
-
 	for t := range ch {
-		s.subtrees[string(t.Prefix)] = t
+		s.subtrees.Store(string(t.Prefix), t)
 		delete(want, string(t.Prefix))
 	}
 
 	// We might not have got all the subtrees we requested, if they don't already exist.
-	// Create empty subtrees for anything left over. As an additional sanity check we refuse
-	// to overwrite anything already in the cache as we determined above that these subtrees
-	// should not exist in the subtree cache map.
+	// Create empty subtrees for anything left over. Note that multiple parallel readers
+	// may be be running this code and touch the same keys, although this doesn't happen
+	// normally.
 	for _, id := range want {
 		prefixLen := id.PrefixLenBits / depthQuantum
 		px := id.Path[:prefixLen]
 		pxKey := string(px)
-		_, exists := s.subtrees[pxKey]
-		if exists {
-			return fmt.Errorf("preload tried to clobber existing subtree for: %v", id)
-		}
-		s.subtrees[pxKey] = s.newEmptySubtree(id, px)
+		s.subtrees.Store(pxKey, s.newEmptySubtree(id, px))
```

Comment:
Should/could this use `LoadOrStore()` to keep the error-checking behaviour looking for overwrites (if it's now permissible for overwrites, we could check that the stored subtree is identical to the one we were about to store)?

Reply:
Yup, good idea. Done.
We should probably do the same for non-empty subtrees `Store` call above, but in a separate PR.

Reply:
Actually, done it all in this PR.